File: 61.rotate-list.java
/*
 * @lc app=leetcode id=61 lang=java
 *
 * [61] Rotate List
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null) return head;
        int cnt = k, len = 0;
        ListNode kHead = head, kTail = head;
        while (kTail != null && cnt-- > 0) {
            kTail = kTail.next;
            len++;
        }
        if (cnt != -1) {
            cnt = k % len;
            kTail = head;
            while (kTail != null && cnt-- > 0) kTail = kTail.next;
        }
        while (kTail.next != null) {
            kHead = kHead.next;
            kTail = kTail.next;
        }
        kTail.next = head;
        head = kHead.next;
        kHead.next = null;
        return head;
    }
}

File: go/61.rotate-list.go
/*
 * @lc app=leetcode id=61 lang=golang
 *
 * [61] Rotate List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func rotateRight(head *ListNode, k int) *ListNode {
    if head == nil || head.Next == nil {
		return head
	}
	len := 1
	for p := head; p.Next != nil; {
		len, p = len+1, p.Next
	}
	if len <= k {
		k = k % len
	}
	if k == 0 {
		return head
	}
	ahead := &ListNode{0, head}
	slow, fast := head, head
	for i := 0; i < k; i++ {
		fast = fast.Next
	}
	for fast.Next != nil {
		slow, fast = slow.Next, fast.Next
	}
	fast.Next = ahead.Next
	ahead.Next = slow.Next
	slow.Next = nil
	return ahead.Next
}
// @lc code=end

